modules/apps/asset/asset-list-service/src/main/java/com/liferay/asset/list/internal/asset/entry/provider/AssetListAssetEntryProviderImpl.java

```diff
@@ -317,37 +320,39 @@ private long[] _filterAssetCategoryIds(long[] assetCategoryIds) {
 		return ArrayUtil.toArray(assetCategoryIdsList.toArray(new Long[0]));
 	}
 
-	private List<AssetListEntryAssetEntryRel>
-		_filterAssetListEntryAssetEntryRels(
-			List<AssetListEntryAssetEntryRel> assetListEntryAssetEntryRels) {
+	private Map<AssetListEntryAssetEntryRel, AssetEntry>
+		_filterAssetListEntryAssetEntryRelAssetEntryMap(
+			Map<AssetListEntryAssetEntryRel, AssetEntry>
+				assetListEntryAssetEntryRelAssetEntryMap) {
 
-		return ListUtil.filter(
-			assetListEntryAssetEntryRels,
-			assetListEntryAssetEntryRel -> {
-				AssetEntry assetEntry = _assetEntryLocalService.fetchEntry(
-					assetListEntryAssetEntryRel.getAssetEntryId());
+		for (Map.Entry<AssetListEntryAssetEntryRel, AssetEntry> entry :
+				assetListEntryAssetEntryRelAssetEntryMap.entrySet()) {
 
-				if ((assetEntry == null) || !assetEntry.isVisible()) {
-					return false;
-				}
+			AssetEntry assetEntry = entry.getValue();
 
-				AssetRendererFactory<?> assetRendererFactory =
-					AssetRendererFactoryRegistryUtil.
-						getAssetRendererFactoryByClassName(
-							assetEntry.getClassName());
+			if ((assetEntry == null) || !assetEntry.isVisible()) {
```

Comment:
Move this null/visible check to ln 468 and don't even add "bad" entries to the map.

Reply:
Same with the null AssetRendererFactory check.